gitpod-io/gitpod
# [ssh-proxy] only send heartbeat for pty sessions

## Description


VS Code Remote SSH will reestablish a ssh connection even if window is not used. It will retrigger heartbeat over and over in SSH Gateway. SSH Gateway should only trigger normal heartbeat if pty sessions is present. See internal chat [[1](https://gitpod.slack.com/archives/C01KGM9BH54/p1666338697815499)]

This PR we add a flag `first` for `ws-manager` `MarkActive` API, if this flag set to `true` it means `ws-manager` should mark active only if the user is never active before. 

## Related Issue(s)

Fixes #

## How to test

1. start a workspace in preview environment, once you see loading page, close the tab, and start a new tab in dashboard(this prevent `supervisor-frontend` send heartbeat)
2. use `kubectl get pods` to get your workspace instanceID
3. use `gpctl workspaces last-heartbeat instanceId` to check your workspace last heartbeat time is empty
4. use copy `copy/paste` ssh command
5. run command like this
```shell
ssh workspaceId#ownerToken@workspaceId.ssh.ws.ak-ssh-gat413da7b457.preview.gitpod-dev.com bash -c "echo aaa > test.txt"
```
6. use `gpctl workspaces last-heartbeat instanceId` to check your workspace last heartbeat time is correct
7. rerun step  5 many times, and each time use step 6 to check workspace last heartbeat is never change
8. use vscode browser to open this workspace, and use step 6 to check last heartbeat is change

## Release Notes

```release-note
Don't trigger heartbeat on all ssh connections. Only for pty sessions.
```

## Documentation


## Werft options:

      If enabled this will build `install/preview`
      Valid options are `all`, `workspace`, `webapp`, `ide`
